file.go

```diff
@@ -1,20 +1,135 @@
 package s3fs
 
 import (
+	"errors"
+	"fmt"
 	"io"
 	"io/fs"
+	"io/ioutil"
 	"path"
 	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3iface"
 )
 
 var (
 	_ fs.File     = (*file)(nil)
 	_ fs.FileInfo = (*fileInfo)(nil)
+	_ io.Seeker   = (*file)(nil)
 )
 
 type file struct {
+	cl     s3iface.S3API
+	bucket string
+	name   string
+
 	io.ReadCloser
-	stat func() (fs.FileInfo, error)
+	stat   func() (fs.FileInfo, error)
+	offset int64
+}
+
+func openFile(cl s3iface.S3API, bucket string, name string) (fs.File, error) {
+	out, err := cl.GetObject(&s3.GetObjectInput{
+		Key:    &name,
+		Bucket: &bucket,
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	statFunc := func() (fs.FileInfo, error) {
+		return stat(cl, bucket, name)
+	}
+
+	if out.ContentLength != nil && out.LastModified != nil {
+		// if we got all the information from GetObjectOutput
+		// then we can cache fileinfo instead of making
+		// another call in case Stat is called.
+		statFunc = func() (fs.FileInfo, error) {
+			return &fileInfo{
+				name:    path.Base(name),
+				size:    *out.ContentLength,
+				modTime: *out.LastModified,
+			}, nil
+		}
+	}
+
+	return &file{
+		cl:         cl,
+		bucket:     bucket,
+		name:       name,
+		ReadCloser: out.Body,
+		stat:       statFunc,
+		offset:     0,
+	}, nil
+}
+
+func (f *file) Read(p []byte) (int, error) {
+	n, err := f.Read(p)
+	f.offset += int64(n)
+	return n, err
+}
+
+func (f *file) Seek(offset int64, whence int) (int64, error) {
+	newOffset := f.offset
+
+	stat, err := f.Stat()
+	if err != nil {
+		return 0, err
+	}
+	size := stat.Size()
+
+	switch whence {
+	case io.SeekStart:
+		newOffset = offset
+	case io.SeekCurrent:
+		newOffset += offset
+	case io.SeekEnd:
+		newOffset = size + offset
+	default:
+		return f.offset, errors.New("s3fs.file.Seek: invalid whence")
+	}
+
+	// If the position has not moved, there is no need to make a new query
+	if f.offset == newOffset {
+		return newOffset, nil
+	}
+
+	if newOffset < 0 {
+		return f.offset, errors.New("s3fs.file.Seek: seeked to a negative position")
+	}
+	if newOffset > size {
+		return f.offset, errors.New("s3fs.file.Seek: seeked to a position exceeding size")
+	}
+	if newOffset == size {
+		f.ReadCloser = ioutil.NopCloser(eofReader{})
+		f.offset = newOffset
+		return f.offset, nil
+	}
+
+	err = f.Close()
```

Comment:
1. should f.Close() be called **after** we successfully got the new object?

2. assuming that s3 client uses regular http.Response, there is something to think about - according to http.Response's documentation:
```
        // The default HTTP client's Transport may not
	// reuse HTTP/1.x "keep-alive" TCP connections if the Body is
	// not read to completion and closed.
```
we definitely can't read the entire body if the file is large, but maybe we should determine if the remaining body is small enough and read it to ioutil.Discard to release the connection

Reply:
1. Changed, closing now happens before any operations that would assign a new value to f.ReadCloser.
2. I'm not quite sure how to implement this. What is small enough? 100 bytes? 1kB?